src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy

```diff
@@ -330,8 +332,41 @@ public abstract class GenerateProtoTask extends DefaultTask {
   @Internal("Input captured by getAlternativePaths(), this is used to query alternative path by locator name.")
   abstract MapProperty<String, FileCollection> getLocatorToAlternativePathsMapping()
 
-  @InputFiles
-  @PathSensitive(PathSensitivity.NONE)
+  @Input
+  @Optional
+  Provider<String> getProtocArtifact() {
+    // Ideally, this value could be mapped from the protocLocator property, but Provider#map cannot return null in
+    // Gradle < 6.2. See https://github.com/gradle/gradle/issues/11979
+    def provider = providerFactory.provider {
+      def protocArtifact = protocLocator.get().artifact
+      if (protocArtifact) {
+        def (groupId, artifact, version) = ToolsLocator.artifactParts(protocArtifact)
+        "$groupId:$artifact:$version".toString()
+      } else {
+        null
+      }
+    }
+
+    outputs.upToDateWhen {
+      !provider.getOrElse("").endsWith("-SNAPSHOT")
```

Comment:
Couldn't decide if it was worth trying to cache on snapshots. A typical plugin consumer isn't going to utilize them.

Reply:
this eagerly resolves the provider when this function is called in order to add an output upToDate condition. i don't think this is a good practice.